stellar-contract-env-host/src/events.rs

```diff
@@ -1,39 +1,227 @@
 #[derive(Clone, Debug)]
-pub(crate) enum HostEvent {
+pub enum HostEvent {
+    #[allow(dead_code)]
     Contract(/*ContractEvent*/),
-    // Same size as LargeDebug but doesn't chew up a heap allocation.
-    SmallDebug {
-        msg: &'static str,
-        // ArrayVec<RawVal, 2> is 3 words: RawVal, RawVal, len.
-        args: ArrayVec<RawVal, 2>,
-    },
-    // Arbitrary set of k/v arguments.
-    LargeDebug {
-        msg: &'static str,
-        // Vec is 3 words: ptr, len, capacity.
-        args: Vec<RawVal>,
-    },
+    Debug(DebugEvent),
 }
 
 #[derive(Clone, Debug, Default)]
 pub struct Events(Vec<HostEvent>);
 
 impl Events {
-    pub(crate) fn add_debug_event(&mut self, msg: &'static str, in_args: &[RawVal]) {
-        let debug = if in_args.len() <= 2 {
-            let args: ArrayVec<RawVal, 2> = ArrayVec::from_iter(in_args.iter().cloned());
-            HostEvent::SmallDebug { msg, args }
-        } else {
-            let args = in_args.into();
-            HostEvent::LargeDebug { msg, args }
+    // Records the smallest variant of a debug HostEvent it can, returning the size of the
+    // in_args slice (for charging to a budget).
+    pub fn record_debug_event(&mut self, de: DebugEvent) -> u64 {
+        let len = de.args.len();
+        self.0.push(HostEvent::Debug(de));
+        len as u64
+    }
+
+    pub fn dump_to_debug_log(&self) {
+        for e in self.0.iter() {
+            match e {
+                HostEvent::Contract() => debug!("Contract event: <TBD>"),
+                HostEvent::Debug(e) => debug!("Debug event: {:?}", e),
+            }
+        }
+    }
+}
+
+/// A cheap record type to store in the events buffer for diagnostic reporting
+/// when something goes wrong. Should cost very little even when enabled. See
+/// [host::debug_event] for normal use.
+#[derive(Clone, Debug)]
+pub struct DebugEvent {
+    pub msg: &'static str,
+    pub args: TinyVec<[RawVal; 2]>,
+}
+
+impl DebugEvent {
+    pub fn new() -> Self {
+        Self {
+            msg: "",
+            args: Default::default(),
+        }
+    }
+
+    pub fn msg(mut self, msg: &'static str) -> Self {
+        self.msg = msg;
+        self
+    }
+
+    pub fn arg<T>(mut self, arg: T) -> Self
+    where
+        RawVal: From<T>,
+    {
+        self.args.push(arg.into());
+        self
+    }
+}
+
+/// Combines a [DebugEvent] with a [Status] that created it, typically
+/// used as a transient type when recording a (possibly enriched)
+/// debug event for a status and then converting the status to a
+/// HostError. See [host::err] for normal use.
+#[derive(Clone, Debug)]
+pub struct DebugError {
+    pub event: DebugEvent,
+    pub status: Status,
+}
+
+impl DebugError {
+    pub fn new<T>(status: T) -> Self
+    where
+        Status: From<T>,
+    {
+        let status: Status = status.into();
+        Self {
+            event: DebugEvent::new().msg("status").arg::<Status>(status.into()),
```

Comment:
This is cool, that you are chaining these calls, each of which returns a `self`. 
Just curious, is there any advantage of doing it this way versus constructing the object by passing in fields directly? 
`event: DebugEvent {msg: "status", args: [status]}`?

Reply:
Just brevity and flexibility at the call site (i.e. you can omit msg or args, or only pass 1 arg, with less noise). This pattern ("builders") is fairly common in Rust because Rust does not support keyword arguments, optional/default arguments, or argument-list overloading of constructors. It's the only real way to ship "flexible construction".

Reply:
I think I'd rather the simpler `DebugEvent { msg: "...", args: tiny_vec!(1, 2, 3) }` as well. It actually appears to be more brief than the longer syntax here.

I'm not opposed to the builder pattern but it is usually beneficial when a type's constraints can't be enforced naturally by the type system, which isn't the case here.